feat: Dashboard

adityahase opened the pull request:
Dashboard is a DocType which contains Dashboard Charts. A Dashboard Chart contains configuration and data source needed to produce a chart.

<details>
<summary>
<b>Dashboard Chart Configuration</b>
</summary>
<img src="https://user-images.githubusercontent.com/8528887/52326390-86252200-2a0e-11e9-83f2-f5793bfd9acb.png"/>
</details>

## Dashboard in action

![pc](https://user-images.githubusercontent.com/8528887/52326704-95f13600-2a0f-11e9-9763-075a21f975dc.gif)

**Note:** Thanks to @netchampfaris

adityahase committed:
feat(dashboard): Create Dashboard page

adityahase committed:
feat(dashboard): Add timespan filter

adityahase committed:
feat(dashboard): Add Time Grain filter

adityahase committed:
feat(dashbard): Set account from filters

adityahase committed:
feat(dashboard): Create chart DocType

adityahase committed:
feat(dashboard): Create Dashboard DocType

adityahase committed:
fix(dashboard): Use valid chart type in chart options

adityahase committed:
fix(dashboard): Do not render single chart by default

adityahase committed:
feat(dashboard): Render multiple charts from charts table

adityahase committed:
fix(dashboard): Change naming for Dashboard DocType

adityahase committed:
feat(dashboard): Set dashboard name as page title

adityahase committed:
feat(dashboard): Add Show Dashboard button

adityahase committed:
feat(dashboard): Translate Show Dashboard button label

adityahase committed:
fix(dashboard): Remove global filters

adityahase committed:
feat(dashboard): Set chart_name as chart title

adityahase committed:
rafactor(dashboard): Use capitalized names in chart type

adityahase committed:
fix(dashboard): Change fieldtype Chart Filters to Code

adityahase committed:
feat(dashboard): Configurable chart width

adityahase committed:
refactor(dashboard): Move DocTypes to Desk module

- Rename Chart to Dashboard Chart

adityahase committed:
refactor(dashboard): Refactor into Class

adityahase committed:
refactor(dashboard): Refactor field names

adityahase committed:
feat(dashboard): Add color and styling

adityahase committed:
feat(code-editor): Add JSON language for Code Editor

adityahase committed:
feat(dashboard): Add dropdown menu to set filters

adityahase committed:
refactor(dashboard): Rename More to Set Filters

adityahase committed:
feat(dashboard): Add Refresh Chart button

adityahase committed:
feat(dashboard): Cache results for Chart

adityahase committed:
feat(dashboard): Convert indentation to tabs

adityahase committed:
feat(dashboard): Update chart after changing filters

adityahase committed:
refactor(dashboard): Abstract away DashboardChart

adityahase committed:
refactor(dashboard): Update last_synced_on on forced refresh

adityahase committed:
fix(dashboard): Add missing semicolons

adityahase committed:
fix(dashboard: Remove empty lines

adityahase committed:
refactor(dashboard): Make Dashboard Chart Form View pretty

adityahase committed:
fix(dashboard): Show currently set values in Dialog

adityahase committed:
feat(dashboard): Show Save Filters button after change in filters

adityahase committed:
style: Linting fixes

adityahase commented:
Please Don't squash

rmehta commented on frappe/core/page/dashboard/dashboard.js:
```diff
@@ -0,0 +1,196 @@
+		});
+	}
+
+	prepare_container() {
```
make this a smaller function

rmehta commented on frappe/core/page/dashboard/dashboard.py:
```diff
@@ -0,0 +1,32 @@
+
+
+def generate_and_cache_results(chart, filters, cache_key):
+	results = frappe.get_attr(chart.method_path)(filters=frappe._dict(filters))
```
Move this to hooks (this is user input)

adityahase committed:
feat(dashboard): Create Dashboard Chart Source doctype

adityahase committed:
feat(dashboard): Show filters in a table

adityahase committed:
refactor(dashboard): Fetch filter fields from .js file

adityahase committed:
feat(dashboard): Create Dashboard Chart Source docs from exported files

adityahase committed:
feat(dashboard): New cache_source decorator for caching

adityahase committed:
fix(dashboard): Fetch data with correct filters

adityahase committed:
refactor(dashboard): Remove tabular filter view

adityahase committed:
fix(dashboard): Remove unnecessary column break

adityahase committed:
refactor(dashboard): Break down prepare_container into smaller functions

adityahase committed:
fix(dashboard): Update error message

adityahase committed:
fix(dashboard): Show update filters modal correctly

adityahase committed:
style: Linting fixes

adityahase committed:
feat(dashboard): Provide modal to update filters

adityahase committed:
feat(dashboard): Use xIsSeries chart option for time series

adityahase committed:
feat(dashboard): Show links to other dashboards in a dropdown menu

Thunderbottom commented:
@frappe-bot create instance

Thunderbottom commented:
@frappe-bot create instance

Thunderbottom commented:
@frappe-bot create instance

Thunderbottom commented:
@frappe-bot create instance

Thunderbottom commented:
@frappe-bot create instance

Thunderbottom commented:
@frappe-bot create instance

Thunderbottom commented:
@frappe-bot create instance

Thunderbottom commented:
@frappe-bot create instance

Thunderbottom commented:
@frappe-bot create instance

Thunderbottom commented:
@frappe-bot create instance

Thunderbottom commented:
@frappe-bot create instance

Thunderbottom commented:
@frappe-bot create instance ['erpnext#16411', 's-2vcpu-2gb']

Thunderbottom commented:
@frappe-bot create instance ['erpnext#16411', 's-2vcpu-2gb']

surajshetty3416 commented:
@frappe-bot create instance

adityahase committed:
fix(dashboard): Use Dashboard Chart Link for listing charts

ccfiel commented:
any updates on this? :)

adityahase committed:
fix(dashboard): Use config insteaad of get_script

Earlier implementation used a get_script function which
served a JS configuration file which was then evaluated
inside the dom.

Now dashboard chart source contains a code field for
storing configuration in JSON

adityahase committed:
style: Linting fixes

adityahase committed:
fix(dashboard): Clear container when rendering another dashboard

adityahase committed:
style: Linting Fixes

adityahase committed:
style: Linting fixes

adityahase commented:
https://github.com/frappe/erpnext_com/pull/292
